### PR TITLE
test: pin AttentionResidualMixer against an independent mix (PRPUNDIT-22)

### PR DESCRIPTION
**Pull request by jiagaoxiang**
Same-repo refile of #1092 so Primus-CI-TAS runs against `AMD-AGI/Primus` (fork PRs do not receive Docker Hub credentials, so `build-docker` / torch unit tests never ran).

This PR was created by dougljia via Test Gap Resolver.

## Summary
- Adds `tests/unit_tests/backends/megatron/kimi_k3/test_attention_residual.py`, the file the module docstring already claimed existed.
- Independent candidate-by-candidate RMS-score / unnormalised-value oracle (not a copy of the production eager kernel), plus `num_blocks=0` identity, hidden-size `ValueError`, dtype cast-back, and `gradcheck`.

Closes test gap **PRPUNDIT-22**.

## Test plan
- [ ] CI with `megatron.core`: `pytest tests/unit_tests/backends/megatron/kimi_k3/test_attention_residual.py` (skipped on this host; `megatron.core` not importable)
- [x] `black --check` / `isort --profile black` on the new files

### CODE REVIEW

File: tests/unit_tests/backends/megatron/kimi_k3/test_attention_residual.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    block_residual = torch.randn(2, 2, hidden_size, dtype=torch.float64, requires_grad=True)
+    mixer.norm_weight.requires_grad_(True)
+    mixer.proj_weight.requires_grad_(True)
+
+    # norm_weight / proj_weight are passed through so gradcheck perturbs the
+    # exact tensors the mixer reads (they are the same objects, not copies).
+    assert torch.autograd.gradcheck(
+        lambda ps, br, nw, pw: mixer(ps, br),
+        (prefix_sum, block_residual, mixer.norm_weight, mixer.proj_weight),
+        eps=1e-6,
```
Comment (jiagaoxiang) [inline tests/unit_tests/backends/megatron/kimi_k3/test_attention_residual.py:228]: **TestIntel PR Steward:** thanks for flagging this — checked it carefully but leaving as is.

`nw`/`pw` are passed into the lambda by object identity, not copied: `mixer.norm_weight.requires_grad_(True)` / `mixer.proj_weight.requires_grad_(True)` are called on the mixer's actual parameter tensors, and those same tensor objects are what's handed to `gradcheck` as `nw`/`pw`. `torch.autograd.gradcheck` perturbs each input's underlying storage in place (via `tensor.data`) and computes the analytical gradient with `autograd.grad(output, inputs)` keyed on object identity — not on how the closure's parameter names are used inside the lambda body. Since `nw is mixer.norm_weight` and `pw is mixer.proj_weight`, every perturbation of `nw`/`pw` mutates the exact tensors `mixer(ps, br)` reads on the next call, so gradients w.r.t. the mixer's own parameters genuinely are exercised by both the numerical and analytical passes, even though the lambda body never references `nw`/`pw` by name.

This is a known idiom for gradchecking a module's internal parameters when they aren't part of the plain function's positional signature (can't gradcheck `module.forward` directly since gradcheck needs a bare function + a flat tuple of "inputs" whose grad it can check by reference). The adjacent comment (two lines up) documents this intentionally. Not changing it — happy to add a defensive `assert nw is mixer.norm_weight and pw is mixer.proj_weight` if you'd like insurance against a future refactor silently breaking the aliasing, but the test itself is correct today.

**Review by jiagaoxiang — COMMENTED** [review]
Response to Copilot's gradcheck finding inline — see thread on tests/unit_tests/backends/megatron/kimi_k3/test_attention_residual.py.

**Comment by jiagaoxiang** [comment]
**TestIntel PR Steward:** follow-up on this poll.

**Copilot review (gradcheck comment):** replied in-thread. Checked it against PyTorch's actual `gradcheck` semantics — `nw`/`pw` in the lambda are the *same objects* as `mixer.norm_weight`/`mixer.proj_weight` (assigned by reference, `requires_grad_(True)` called on the real attributes), and `gradcheck` perturbs/differentiates by tensor identity, not by how the lambda body names its parameters. So gradients w.r.t. the mixer's own parameters are genuinely exercised despite the lambda ignoring `nw`/`pw` by name. Not changing the test; happy to add a defensive identity assert if a reviewer wants extra insurance.

**CI:** the `build-docker (3.12)` failure surfaced by this poll is from the superseded fork-context run of the same commit (the original PR #1092 from `jiagaoxiang/Primus`, which — as noted in this PR's description — never got Docker Hub credentials as a fork PR). That's not this PR's own CI: the same-repo run for this exact head SHA in `AMD-AGI/Primus` (run [33895332771](https://github.com/AMD-AGI/Primus/actions/runs/33895332771)) shows `dependency-review` ✅, `code-lint (3.12)` ✅, and `build-docker (3.12)` ✅ (succeeded once it ran against this repo). `run-unittest-jax` and `run-unittest-torch` are still queued on that same run — nothing broken, just waiting on runners. No push needed right now.

Will keep watching for the jax/torch results, further review comments, and a human approval before merge.

**Review by Xiaoming-AMD — APPROVED** [review]

**Review by Xiaoming-AMD — APPROVED** [review]
